docs/architecture.md

```diff
@@ -0,0 +1,545 @@
+
+### data flow
+
+https://mermaid.js.org/syntax/sequenceDiagram.html
+
+```mermaid
+---
+title: request flow
+---
+flowchart LR
+    user_defined_types
+    odata_request1[odata_request]
+    http_request1[http_request]
+    http_client
+    http_handler
+    http_request2[http_request]
+    odata_request2[odata_request]
+    data_store_request
+
+    user_defined_types --> odata_request1
+    odata_request1 --> http_request1
+    http_request1 --> http_client
+    http_client --> http_handler
+    http_handler --> http_request2
+    http_request2 --> odata_request2
+    odata_request2 --> data_store_request
+
+    style odata_request1 fill:#08FF08
+    style odata_request2 fill:#08FF08
+    style http_request1 fill:#FBF719
+    style http_request2 fill:#FBF719
+```
+
+```mermaid
+---
+title: response flow
+---
+flowchart LR
+    data_store_response
+    odata_response1[odata_response]
+    http_response1[http_response]
+    http_handler
+    http_client
+    http_response2[http_response]
+    odata_response2[odata_response]
+    user_defined_types
+
+    data_store_response --> odata_response1
+    odata_response1 --> http_response1
+    http_response1 --> http_handler
+    http_handler --> http_client
+    http_client --> http_response2
+    http_response2 --> odata_response2
+    odata_response2 --> user_defined_types
+
+    style odata_response1 fill:#08FF08
+    style odata_response2 fill:#08FF08
+    style http_response1 fill:#FBF719
+    style http_response2 fill:#FBF719
+```
+
+Notice that in each case, the request and response types are the same for both the client and the service. There are three side effects of this:
+1. The entire HTTP portion of the flow can be removed, meaning that **any** routing infrastructure can be leveraged while sharing the same common "odata_request" and "odata_response" types. This would allow ASP.NET services and azure functions to share the same basic scaffolding.
+2. Because the HTTP portion is not strictly necessary for processing a request, it means that odata services can be easily composed at this layer. If an odata_request is sent by a client using the same type that it is processed by the service, then it means that additional odata services can themselves be treated as "data stores". More on this [later](#a-common-abstraction-for-handling-a-request).
+3. These types can be shared by consumers of the odata libraries, but currently they are not. To see the impact of this sharing, let's dive into the mechanics of each translation of the data:
+
+#### user defined types to odata request
+
+```mermaid
+sequenceDiagram
+    participant user defined types
+    participant odata request AST
+
+    user defined types->>odata request AST: serialized to
+```
+
+#### odata request to http request
+
+```mermaid
+sequenceDiagram
+    participant odata request AST
```

Comment:
Did I get these backwards? Which comes first, the AST, or the CST, when the input is the http request?

Reply:
Currently, we are calling the CST "syntactic ast" and the AST "semantic ast"

Reply:
@mikepizzo I am not finding any sources that refer to these concepts by the names "syntactic AST" and "semantic AST", but I am finding lots of literature that uses "AST" and either "CST" or "parse tree". I'm happy to rename if we really need to, but I'd like to normalize on established names.

Reply:
After some more discussion, my previous comments were incorrect. What we have today is a "syntactic AST" that follows a different grammar than the OData ABNF. This grammar is context-free and allows preliminary parsing to happen without the need of the EDM model. The "semantic AST" follows the OData ABNF with the context of the EDM model. 

I will include a note in the follow-ups that there **must** be a concept of this "syntactic AST" to maintain parity.